routers/user/auth.go

```diff
@@ -664,8 +664,23 @@ func LinkAccount(ctx *context.Context) {
 		return
 	}
 
-	ctx.Data["user_name"] = gothUser.(goth.User).NickName
-	ctx.Data["email"] = gothUser.(goth.User).Email
+	uname := gothUser.(goth.User).NickName
+	email := gothUser.(goth.User).Email
+	ctx.Data["user_name"] = uname
+	ctx.Data["email"] = email
+
+	if "" != uname {
+		u, _ := models.GetUserByName(uname)
+		if u != nil {
+			ctx.Data["user_name_exists"] = "true"
```

Comment:
Probably you can use same ctx variable in both places, ex. `ctx.Data["user_already_exists"] = true`

Reply:
True for this PR, but it’s intentional because the UX is different in the case that it’s one or the other.

Reply:
True for this PR, but it’s intentional because the UX is different in the case that it’s one or the other.

Reply:
Let's say that my favorite username `aj` is available on gitea, but I had to settle for my second favorite `coolaj86` on github.

Let's say that I trust github less so I used coolaj86+github@gmail.com, but on gitea I simply use coolaj86@gmail.com.

If either match I should be shown the "Link Account" flow by default, but the value that I show in the username or email field will be the value that matches, not the value that doesn't match.

Reply:
That doesn't impose the necessity for the two different ctx variables.
You can just change the `ctx.Data["user_name"]` to whatever's the matching one. No need to differ between the two cases.

Reply:
Yes, but I expect I'd then add something like `ctx.Data["user_type"] = "email"` so that I can use it in templates later on. So "six of one, half a dozen of the other".

Reply:
IMO we shouldn't introduce variables that we might need in another PR.
If a PR needs it you should introduce it with that PR, otherwise KISS.

Reply:
I get that.

Bear in mind that I'm already going to be doing 2x-3x the amount of work that I would need to do to get this working for myself. The reason I'm going through this process, even though step by step, plus reviews, etc, it'll take significantly longer than the oneshot that I'd like to do, is because I really believe in gitea and I don't just want my instance to be better, I want everyone's instance to be better.

I'm on a path and my choices aren't arbitrary or purely hypothetical.

Being the new guy I also recognize that I haven't earned that trust yet, so I'll go through the motions.

And all that said, I probably won't get through everything this week or this month - certainly not going the route that will result in everyone benefitting from the changes. To that end, perhaps it is best to reduce it just because it's possible someone else will come in inbetween now and then...

Thanks for the thorough review. I'm out of time for the night, but I hope to be back tomorrow with the suggested edits.

:)